judge/judgeapi.py

```diff
@@ -100,6 +104,65 @@ def judge_submission(submission, rejudge=False, batch_rejudge=False, judge_id=No
     return success
 
 
+def batch_judge_submission(submissions, rejudge=False, judge_id=None):
+    from .models import Submission, SubmissionTestCase
+    updates = {
+        'time': None, 'memory': None, 'points': None, 'result': None, 'case_points': 0, 'case_total': 0,
+        'error': None, 'rejudged_date': timezone.now() if rejudge else None, 'status': 'QU',
+        'is_pretested': Coalesce(
+            Subquery(
+                Submission.objects.filter(pk=OuterRef('pk'))
+                .annotate(pretested=F('contest_object__run_pretests_only').bitand(F('contest__problem__is_pretested')))
+                .values_list('pretested')[:1],
+                output_field=BooleanField(),
+            ),
+            False,
+        ),
+    }
+
+    with transaction.atomic():
+        submission_queryset = Submission.objects.filter(id__in=map(attrgetter('id'), submissions)) \
```

Comment:
If the argument is a queryset, then you wouldn't need to do this hack and could just do `.exclude` on the queryset directly. And then you can just load this queryset to get the `submissions` instead of filtering in Python.

Reply:
If the argument is a queryset, that means it will be lazy. We would be doing the splicing at the database layer, so the slice may change between batches. For example, consider we were trying to rejudge `Submission.objects.exclude(status='QU')`. As we're processing each batch, [0:batch_size] would always contain a not judged queryset since the previous batch's submissions would become `QU` (and thus filtered out). We would only process the [0:batch_size] splice once, and so we would be missing some submissions.

From what I could tell, there's 2 solutions:
 - Use a single batch. This wouldn't handle large querysets well, so I opted not to do this.
 - Evaluate the queryset so we have a fixed set of submissions to judge in each batch, which is what is currently done via `.iterator()`.

Other solutions are welcome.